### PR TITLE
Support spaces in paths in static web publishing.

### LINKED ISSUE #2725: Web Publishing doesn't serve static files with spaces
Steps I took to reproduce:

* create a Davros grain, add a file with spaces in it.
* turn on web publishing (this currently just symlinks davros's `data` directory to `www`)
* go to the file, attempt to access its public URL.

It appears that it doesn't urldecode the URL paths before serving the files.

(zarvox comment from IRC: _just thinking out loud: it should probably be the supervisor that handles decoding the path before trying to open the file it serves_)

### PR DESCRIPTION
**Pull request by kentonv**
Fixes #2725.

### CODE REVIEW

**Comment by dwrensha** [comment]
This would be a breaking change with at least Hacker CMS, which automatically escapes spaces in filenames.

**Comment by kentonv** [comment]
I guess that's true. I guess the alternative right now is to tell apps to name files using escaped names.

@mnutt Thoughts?

**Comment by mnutt** [comment]
@dwrensha how do those files look on disk and how are they referred to in URLs? They have `%20`s in the filenames? I have files on disk with spaces, and am not able to refer to them at all with URLs.

**Comment by dwrensha** [comment]
On disk, the file names have spaces escaped as "%20" and they can be refered to in URLs with a path like "/my%20file". After the change in this PR, such URLs would need to be double-escaped, like "/my%2520file".

I agree that #2725 is a bug, but I'm worried that some people might be depending on the old behavior.

Maybe we should add a parameter to `HackSessionContext.getPublicId()` to indicate which escaping style is expected by the app.

**Comment by mnutt** [comment]
Hmm, agreed that we don't want to break existing behavior, and adding a new parameter probably makes sense. Another option could be to first look for the file with spaces, and if it doesn't exist then fall back to `%20`-escaped filenames.

**Comment by kentonv** [comment]
OK, I added a commit that falls back to the URL-encoded path if the decoded path doesn't work. Thoughts?

File: shell/server/pre-meteor.js [hunk 1]
```diff
@@ -127,6 +131,13 @@ function wwwHandlerForGrain(grainId) {
           });
           response.end("redirect: /" + path + "/");
         } else if (status === "notFound") {
+          if (path !== encoded) {
+            // Compatibility hack: Try geting the URL-encoded version of the path. We used to not
```
Comment (dwrensha) [inline shell/server/pre-meteor.js]: s/geting/setting/

Comment (kentonv) [inline shell/server/pre-meteor.js]: Actually "getting" is what I intended, as in we're "getting" the file from the grain.

Though I meant it with two t's...

**Comment by mnutt** [comment]
Thanks, this is great!

**Pull request merged**